### PR TITLE
Create and allot LP token on Add liquidity

### PR DESCRIPTION
**Pull request by abhishek-k-sharma-lab49**
**Description**:

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: bin/contracts/AbstractLPToken.abi [hunk 1]
```diff
@@ -0,0 +1 @@
...uint32"},{"internalType":"address","name":"autoRenewAccount","type":"address"},{"internalType":"uint32","name":"autoRenewPeriod","type":"uint32"}],"internalType":"struct IHederaTokenService.Expiry","name":"expiryInfo","type":"tuple"}],"name":"updateTokenExpiryInfo","outputs":[{"internalType":"int256","name":"responseCode","type":"int256"}],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address","name":"token","type":"address"},{"components":[{"internalType":"uint256","name":"keyType","type":"uint256"},{"components":[{"internalType":"bool","name":"inheritAccountKey","type":"bool"},{"internalType":"address","name":"contractId","type":"address"},{"internalType":"bytes","name":"ed25519","type":"bytes"},{"internalType":"bytes","name":"ECDSA_secp256k1","type":"bytes"},{"internalType":"address","name":"delegatableContractId","type":"address"}],"internalType":"struct IHederaTokenService.KeyValue","name":"key","type":"tuple"}],"internalType":"struct IHederaTokenService.TokenKey[]","name":"keys","type":"tuple[]"}],"name":"updateTokenKeys","outputs":[{"internalType":"int64","name":"responseCode","type":"int64"}],"stateMutability":"nonpayable","type":"function"}]
```
Comment (amandeeprehal-lab49) [inline bin/contracts/AbstractLPToken.abi:1]: Why do you need these files?

File: bin/contracts/HederaToken.abi [hunk 2]
```diff
@@ -0,0 +1 @@
+[{"inputs":[],"name":"name","outputs":[{"internalType":"string","name":"","type":"string"}],"stateMutability":"view","type":"function"}]
```
Comment (amandeeprehal-lab49) [inline bin/contracts/HederaToken.abi:1]: Why do you need these files?

File: contracts/AbstractLPToken.sol [hunk 3]
```diff
@@ -0,0 +1,61 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractLPToken.sol:2]: We just need .sol files.

File: contracts/AbstractLPToken.sol [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+pragma experimental ABIEncoderV2;
+
+import "./common/hedera/HederaResponseCodes.sol";
+import "./common/IBaseHTS.sol";
+import "./common/hedera/IHederaTokenService.sol";
+import "./common/hedera/HederaTokenService.sol";
+import "./common/hedera/ExpiryHelper.sol";
+
+abstract contract HederaToken {
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractLPToken.sol:11]: Where are you using `HederaToken`?

File: contracts/AbstractLPToken.sol [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    // Abstract Functions
+    function mintToken(uint64 amount) internal virtual returns (int responseCode, uint64 newTotalSupply, int64[] memory serialNumbers);    
+    function associateTokenInternal(address account,  address _token) internal virtual returns(int);
+    function transferTokenInternal(address _token, address sender, address receiver, int64 amount) internal virtual returns(int);
+
+
+    function initializeParams(address _lpToken, IBaseHTS _tokenService) external {
+         // instantiate the list of keys we'll use for token create
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractLPToken.sol:33]: Pls remove these comments all over. If there is complex logic you can keep comment not for single things.

File: contracts/AbstractLPToken.sol [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        // logic to decide quantity of LP
+        uint64 mintingAmount = sqrt(amountA * amountB);
+        //Associate LP to user
+        associateTokenInternal(_toUser, lpToken);
+        ////mint new amount of LP
+        mintToken(mintingAmount);
+        tokenShare[_toUser] = tokenShare[_toUser] + mintingAmount;
+        // transfer Lp to users account
+        transferTokenInternal(lpToken, address(tokenService), _toUser, int64(mintingAmount));
+        return 22;
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractLPToken.sol:50]: You can return `HederaResponseCodes.SUCCESS`

Comment (amandeeprehal-lab49) [reply contracts/AbstractLPToken.sol:50]: Or define your constant and use across.

File: contracts/AbstractLPToken.sol [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    function sqrt(uint64 x) public pure returns (uint64 y) {
```
Comment (amandeeprehal-lab49) [inline contracts/AbstractLPToken.sol:53]: Can we use a good library for square root?

File: contracts/ILPToken.sol [hunk 8]
```diff
@@ -0,0 +1,10 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import "./common/hedera/HederaTokenService.sol";
```
Comment (amandeeprehal-lab49) [inline contracts/ILPToken.sol:4]: Why you need this - service?

File: contracts/AbstractLPToken.sol (hunk 3 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/AbstractLPToken.sol:2]: @amandeeprehal-lab49 I could not understand, should I remove pragma?

File: contracts/LPToken.sol [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+pragma solidity ^0.8.0;
+
+import "./common/hedera/HederaResponseCodes.sol";
+import "./common/IBaseHTS.sol";
+import "./AbstractLPToken.sol";
+
+contract LPToken is AbstractLPToken {
+    
+    function mintToken(uint64 amount) override internal virtual 
+     returns (int responseCode, uint64 newTotalSupply, int64[] memory serialNumbers) {
```
Comment (amandeeprehal-lab49) [inline contracts/LPToken.sol:11]: why this serialNumbers?

File: contracts/common/BaseHTS.sol [hunk 10]
```diff
@@ -4,7 +4,14 @@ pragma solidity ^0.8.0;
 import "./hedera/HederaTokenService.sol";
 import "./hedera/HederaResponseCodes.sol";
 
+
 contract BaseHTS is HederaTokenService {
+    address internal lpToken;
```
Comment (amandeeprehal-lab49) [inline contracts/common/BaseHTS.sol:9]: This is not required.

File: contracts/common/BaseHTS.sol [hunk 11]
```diff
@@ -4,7 +4,14 @@ pragma solidity ^0.8.0;
...
+
+    function initialize(address _lpToken) public {
```
Comment (amandeeprehal-lab49) [inline contracts/common/BaseHTS.sol:11]: This is not required as this is just a delegate service. It doesn't have state.

File: contracts/common/IBaseHTS.sol [hunk 12]
```diff
@@ -13,4 +13,10 @@ interface IBaseHTS {
 
     function associateTokensPublic(address account, address[] memory tokens) 
         external returns (int responseCode);
+
+    function mintTokenPublic(address token, uint64 amount) external
+        returns (int responseCode, uint64 newTotalSupply, int64[] memory serialNumbers);
```
Comment (amandeeprehal-lab49) [inline contracts/common/IBaseHTS.sol:18]: Again, why we need serialNumbers?

File: deployment/LPToken.ts [hunk 13]
```diff
@@ -0,0 +1,100 @@
+
+import { Deployment } from "./deployContractOnTestnet";
```
Comment (amandeeprehal-lab49) [inline deployment/LPToken.ts:2]: Its not getting used.

File: deployment/LPToken.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+import {TokenCreateTransaction, FileCreateTransaction, FileAppendTransaction, AccountId, PrivateKey,
```
Comment (amandeeprehal-lab49) [inline deployment/LPToken.ts:4]: Pls remove unused imports.

File: deployment/LPToken.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+  ContractCreateTransaction, TokenType, TokenSupplyType, Hbar, Client, ContractId, AccountCreateTransaction, KeyList,
+  ContractUpdateTransaction, ContractInfoQuery, ContractExecuteTransaction,
+  ContractFunctionParameters, TokenUpdateTransaction, TokenInfoQuery, TokenAssociateTransaction, AccountBalanceQuery, TokenId
+} from "@hashgraph/sdk";
+import * as fs from "fs";
+import dotenv from "dotenv";
+
+dotenv.config({ path: '../../.env' });
```
Comment (amandeeprehal-lab49) [inline deployment/LPToken.ts:12]: Passing path parameter is not required.

File: deployment/LPToken.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        .setMaxAutomaticTokenAssociations(20)
+        .setInitialBalance(1)
+        .execute(client);
+
+    let createAccountRx = await createAccountTx.getReceipt(client);
+    const aliceAccount = createAccountRx.accountId;
+    console.log(`- Alice account is ${aliceAccount?.toString()} \n privatekey: ${aliceKey}`);
+}
+
+async function deployTokenContract() {
```
Comment (amandeeprehal-lab49) [inline deployment/LPToken.ts:32]: Pls let us use the one existing `deployContractOnTestnet`. Let us discuss this.

File: deployment/baseContract.ts [hunk 17]
```diff
@@ -1,11 +1,93 @@
...
 
 async function main() {
-    const deployment = new Deployment();
-    const filePath = "./artifacts/contracts/common/BaseHTS.sol/BaseHTS.json";
-    const deployedContract = await deployment.deployContract(filePath, []);
-    console.log("baseContract deployed.");
+    await deployBaseContract()
+}
+
+async function deployBaseContract() {
```
Comment (amandeeprehal-lab49) [inline deployment/baseContract.ts:17]: Same goes for this, let us reuse what we already have.

File: integrationTest/lpToken.ts [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  PrivateKey,
+  TokenId,
+  ContractExecuteTransaction,
+  ContractFunctionParameters,
+  Client,
+  ContractId,
+  AccountBalanceQuery
+} from "@hashgraph/sdk";
+
+const createClient = () => {
```
Comment (amandeeprehal-lab49) [inline integrationTest/lpToken.ts:13]: Pls use ClientManagement class it gives everything.

File: integrationTest/lpToken.ts [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+const client = createClient();
+const tokenA = TokenId.fromString("0.0.47646195").toSolidityAddress();
+let tokenB = TokenId.fromString("0.0.47646196").toSolidityAddress();
+const treasure = AccountId.fromString("0.0.47645191").toSolidityAddress();
+const treasureAccountId = AccountId.fromString("0.0.47645191");
+const treasureKey = PrivateKey.fromString("308ed38983d9d20216d00371e174fe2d475dd32ac1450ffe2edfaab782b32fc5");
+
+const contractId = ContractId.fromString("0.0.48190509"); //6 Sep 04:45
+
+const itegrationTestLPToken = async () => {
```
Comment (amandeeprehal-lab49) [inline integrationTest/lpToken.ts:37]: Wrong spelling

File: integrationTest/lpToken.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+  const aliceBalance1 = await new AccountBalanceQuery()
+      .setAccountId(aliceAccount2)
+      .execute(client);
+
+  console.log(aliceBalance1.tokens);
+  }
+};
+
+async function main() {
+  await itegrationTestLPToken();
```
Comment (amandeeprehal-lab49) [inline integrationTest/lpToken.ts:73]: Wrong spelling

File: integrationTest/swapV2.ts [hunk 21]
```diff
@@ -236,17 +248,19 @@ const getInGivenOut =async () => {
   console.log(`For tokenBQty ${tokenBQty} the getInGivenOut tokenAQty is ${tokenAQty}. \n`);
 };
 
-
 async function main() {
-  await initialize();
+  //await initialize();
```
Comment (amandeeprehal-lab49) [inline integrationTest/swapV2.ts:252]: Why commented this?

File: integrationTest/lpToken.ts [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+    throw new Error(
+      "Environment variables myAccountId and myPrivateKey must be present"
+    );
+  }
+
+  const client = Client.forTestnet();
+  client.setOperator(myAccountId, myPrivateKey);
+  return client;
+};
+
```
Comment (amandeeprehal-lab49) [inline integrationTest/lpToken.ts:27]: Also, how about add functional test in swapV2.test?

File: bin/contracts/AbstractLPToken.abi (hunk 1 above)
Comment (abhishek-k-sharma-lab49) [reply bin/contracts/AbstractLPToken.abi]: removed

File: contracts/AbstractLPToken.sol (hunk 4 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/AbstractLPToken.sol]: removed

File: contracts/AbstractLPToken.sol (hunk 5 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/AbstractLPToken.sol]: removed

File: contracts/AbstractLPToken.sol (hunk 6 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/AbstractLPToken.sol]: done

File: contracts/ILPToken.sol (hunk 8 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/ILPToken.sol]: removed

File: contracts/LPToken.sol (hunk 9 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/LPToken.sol]: removed

File: contracts/common/BaseHTS.sol (hunk 10 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/common/BaseHTS.sol]: removed

File: contracts/common/BaseHTS.sol (hunk 11 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/common/BaseHTS.sol]: removed

File: contracts/common/IBaseHTS.sol (hunk 12 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/common/IBaseHTS.sol]: removed

File: deployment/LPToken.ts (hunk 13 above)
Comment (abhishek-k-sharma-lab49) [reply deployment/LPToken.ts:2]: removed

File: deployment/LPToken.ts (hunk 14 above)
Comment (abhishek-k-sharma-lab49) [reply deployment/LPToken.ts]: removed

File: deployment/LPToken.ts (hunk 15 above)
Comment (abhishek-k-sharma-lab49) [reply deployment/LPToken.ts:11]: removed

File: deployment/baseContract.ts (hunk 17 above)
Comment (abhishek-k-sharma-lab49) [reply deployment/baseContract.ts:16]: not done

File: integrationTest/lpToken.ts (hunk 18 above)
Comment (abhishek-k-sharma-lab49) [reply integrationTest/lpToken.ts:13]: done

File: integrationTest/lpToken.ts (hunk 19 above)
Comment (abhishek-k-sharma-lab49) [reply integrationTest/lpToken.ts]: done

File: integrationTest/lpToken.ts (hunk 20 above)
Comment (abhishek-k-sharma-lab49) [reply integrationTest/lpToken.ts]: done

File: integrationTest/swapV2.ts (hunk 21 above)
Comment (abhishek-k-sharma-lab49) [reply integrationTest/swapV2.ts]: done

File: integrationTest/lpToken.ts (hunk 22 above)
Comment (abhishek-k-sharma-lab49) [reply integrationTest/lpToken.ts:27]: this is for extra testing

File: package-lock.json [hunk 23]
```diff
@@ -15134,7 +15134,9 @@
       "resolved": "https://registry.npmjs.org/ganache-core/-/ganache-core-2.13.2.tgz",
       "integrity": "sha512-tIF5cR+ANQz0+3pHWxHjIwHqFXcVo0Mb+kcsNhglNFALcYo49aQpnS9dqHartqPfMFjiHh/qFoD3mYK0d/qGgw==",
       "bundleDependencies": [
-        "keccak"
+        "keccak",
+        "node-addon-api",
```
Comment (amandeeprehal-lab49) [inline package-lock.json:15138]: Pls double check this.

**Review by amandeeprehal-lab49 — APPROVED** [review]

**Pull request merged**